src/heading.js

```diff
@@ -27,46 +25,120 @@ export default class Heading extends Plugin {
 	 * @inheritDoc
 	 */
 	static get requires() {
-		return [ HeadingEngine ];
+		return [ Paragraph, HeadingEngine ];
 	}
 
 	/**
 	 * @inheritDoc
 	 */
 	init() {
 		const editor = this.editor;
-		const command = editor.commands.get( 'heading' );
-		const options = command.options;
-		const collection = new Collection();
-
-		// Add options to collection.
-		for ( const { id, label } of options ) {
-			collection.add( new Model( {
-				id, label
+		const dropdownItems = new Collection();
+		const options = this._getLocalizedOptions();
+		const commands = [];
+		let defaultOption;
+
+		for ( let option of options ) {
+			// Add the option to the collection.
+			dropdownItems.add( new Model( {
+				modelElement: option.modelElement,
+				label: option.title
 			} ) );
+
+			commands.push( editor.commands.get( option.modelElement ) );
+
+			if ( !defaultOption && option.modelElement == 'paragraph' ) {
+				defaultOption = option;
+			}
 		}
 
 		// Create dropdown model.
 		const dropdownModel = new Model( {
 			withText: true,
-			items: collection
+			items: dropdownItems
 		} );
 
-		// Bind dropdown model to command.
-		dropdownModel.bind( 'isEnabled' ).to( command, 'isEnabled' );
-		dropdownModel.bind( 'label' ).to( command, 'value', option => option.label );
+		dropdownModel.bind( 'isEnabled' ).to(
+			// Bind to #isEnabled of each command...
+			...getCommandsBindingTargets( commands, 'isEnabled' ),
+			// ...and set it true if any command #isEnabled is true.
+			( ...areEnabled ) => areEnabled.some( isEnabled => isEnabled )
+		);
+
+		dropdownModel.bind( 'label' ).to(
+			// Bind to #value of each command...
+			...getCommandsBindingTargets( commands, 'value' ),
+			// ...and chose the title of the first one which #value is true.
+			( ...areActive ) => {
+				const index = areActive.findIndex( value => value );
+
+				// If none of the commands is active, display the first one.
+				return ( options[ index ] || defaultOption ).title;
+			}
+		);
 
 		// Register UI component.
 		editor.ui.componentFactory.add( 'headings', ( locale ) => {
 			const dropdown = createListDropdown( dropdownModel, locale );
 
 			// Execute command when an item from the dropdown is selected.
-			this.listenTo( dropdown, 'execute', ( { source: { id } } ) => {
-				editor.execute( 'heading', { id } );
+			this.listenTo( dropdown, 'execute', ( { source: { modelElement } } ) => {
```

Comment:
The param of the callback is impossible for me to parse. Please use more human-readable solution.